core/src/main/java/org/mskcc/cbio/portal/dao/DaoGenePanel.java

```diff
@@ -181,4 +198,103 @@ public static void deleteGenePanel(GenePanel genePanel) throws DaoException {
+
+        if (stableId == null) {
+            throw new DaoException("Gene Panel stable ID cannot be null.");
+        }
+
+        updateGenePanelGeneList(internalId, incoming);
+    }
+
+    /**
+     * Prints the genes to be added and removed from a gene panel in the database.
+     *
+     * @param genePanel
+     * @param incoming
+     * @throws DaoException
+     */
+    public static void updatePreview(GenePanel genePanel, Set<CanonicalGene> incoming) throws DaoException {
+        Pair pair = getAddRemove(incoming, extractGenePanelGenes(genePanel.getInternalId()));
+        String add = pair.add.stream()
+            .map(CanonicalGene::toString)
+            .collect(Collectors.joining(", "));
+        String remove = pair.remove.stream()
+            .map(CanonicalGene::toString)
+            .collect(Collectors.joining(", "));
+
+        System.out.printf("ADD: %s\n", add);
+        System.out.printf("REMOVE: %s\n", remove);
+    }
+
+    private static boolean genePanelExists(Integer internalId) throws DaoException {
+        Connection con = null;
+        PreparedStatement pstmt = null;
+        ResultSet rs = null;
+        Map<String, GenePanel> genePanelMap;
+
+        try {
+            con = JdbcUtil.getDbConnection(DaoGenePanel.class);
+            pstmt = con.prepareStatement("SELECT * FROM gene_panel WHERE internal_id = ?");
+            pstmt.setInt(1, internalId);
+            rs = pstmt.executeQuery();
+            genePanelMap = extractGenePanelMap(rs);
+            return !genePanelMap.isEmpty();
+        } catch (SQLException e) {
+            return false;
+        } finally {
+            JdbcUtil.closeAll(DaoGenePanel.class, con, pstmt, rs);
+        }
+    }
+
+    /**
+     * Inserts and deletes genes from a gene panel represented by an internalId in
+     * the database.
+     */
+    private static void updateGenePanelGeneList(Integer internalId, Set<CanonicalGene> incoming) throws DaoException {
+        Connection con = null;
+        PreparedStatement pstmt = null;
+        ResultSet rs = null;
+
+        try {
+            if (!genePanelExists(internalId)) {
+                System.out.println("Gene panel does not exist in gene_panel, exiting");
+                return;
+            }
+
+            con = JdbcUtil.getDbConnection(DaoGenePanel.class);
+            Pair pair = getAddRemove(incoming, extractGenePanelGenes(internalId));
+            Set<CanonicalGene> toAdd = pair.add;
+            Set<CanonicalGene> toRemove = pair.remove;
+
+            // Add and remove genes from specified gene panel
+            for (CanonicalGene canonicalGene : toAdd) {
+                pstmt = con.prepareStatement("INSERT INTO gene_panel_list (`INTERNAL_ID`, `GENE_ID`) VALUES (?,?)");
+                pstmt.setInt(1, internalId);
+                pstmt.setLong(2, canonicalGene.getEntrezGeneId());
+                pstmt.executeUpdate();
+            }
+            for (CanonicalGene canonicalGene : toRemove) {
```

Comment:
It looks like its possible to remove all genes from the panel.  Should we check this and if so, should we prevent this from happening?

Reply:
In UpdateGenePanel.java, the user prompt in importData() verifies the changes that will be made to the gene panel, like so:
![image](https://user-images.githubusercontent.com/33106214/88946824-b7765b00-d25d-11ea-9cb9-2b14761edf38.png)
Other than that, I asked Luke if importing an empty gene panel should allow for the removal of all existing genes, and he stated that behavior sounds reasonable. I suppose if the user wanted to add back the genes, they could pull up an older version of their input file with the gene names. 
I'll ping Luke about whether after all genes are removed, that the gene panel should be removed from gene_panel_list as well

Reply:
ah, so removal of all genes means stop the update script and don't make the deletions

Reply:
It's not meaningful to have a panel in the database, which samples refer to in sample_profile, that contain no genes.  I'm also not sure what the affect is on the web page.  One approach is to remove the the panel and subsequent records in sample_profile (via cascade), but thats a bit messy and could happen by mistake.

Reply:
Spoke with @Luke-Sikina via Slack.  Best approach is not to prevent an update which results in empty panel in database.

Reply:
Oh, does that mean that within UpdateGenePanel, checking if the incoming set of genes is empty and aborting the script is not needed?

Reply:
I've just added the check on lines 77-80 on UpdateGenePanel, after the gene panel null check:
```java
        if (genePanel == null) {
            ProgressMonitor.logWarning("Gene panel " + stableId + " does not exist in the database! Exiting.");
            return;
        }

        if (canonicalGenes == null || canonicalGenes.isEmpty()) {
            System.err.println("Incoming gene panel is empty, which would result in the removal of all genes from gene panel " + stableId + ". Exiting.");
            return;
        }
```

Reply:
Yeah, checking/aborting when the incoming gene list from the panel file is empty should suffice, thanks for the adjustment.

Reply:
Thank you for the reviews!

Reply:
I've just updated the markdown file again to reflect that trying to import an empty gene panel will abort the script.